consensus.go

```diff
@@ -0,0 +1,235 @@
+	"fmt"
+	"sort"
+	"time"
+
+	"decred.org/dcrwallet/v2/errors"
+	w "decred.org/dcrwallet/v2/wallet"
+	"github.com/decred/dcrd/chaincfg/chainhash"
+	"github.com/decred/dcrd/chaincfg/v3"
+)
+
+// GetVoteChoices returns configured vote preferences for each agenda of the
+// latest supported stake version for the wallet with the specified walletID.
+func (mw *MultiWallet) GetVoteChoices(hash string, walletID int) (*GetVoteChoicesResult, error) {
+	wallet := mw.WalletWithID(walletID)
+	if wallet == nil {
+		return nil, errors.New(ErrNotExist)
+	}
+	wal := wallet.Internal()
+
+	var ticketHash *chainhash.Hash
+	if hash != "" {
+		hash, err := chainhash.NewHashFromStr(hash)
+		if err != nil {
+			return nil, fmt.Errorf("inavlid hash: %w", err)
+		}
+		ticketHash = hash
+	}
+
+	version, agendas := getAllAgendas(wal.ChainParams())
+	resp := &GetVoteChoicesResult{
+		Version: version,
+		Choices: make([]AgendaVoteChoice, len(agendas)),
+	}
+
+	ctx := wallet.shutdownContext()
+	choices, _, err := wal.AgendaChoices(ctx, ticketHash)
+	if err != nil {
+		return nil, err
+	}
+
+	for i := range choices {
+		agenda := mw.agenda(agendas, choices[i].AgendaID)
+		resp.Choices[i] = AgendaVoteChoice{
+			AgendaID:          choices[i].AgendaID,
+			AgendaDescription: agenda.Vote.Description,
+			ChoiceID:          choices[i].ChoiceID,
+			ChoiceDescription: "", // Set below
+		}
+		for _, choice := range agenda.Vote.Choices {
+			if choices[i].ChoiceID == choice.Id {
+				resp.Choices[i].ChoiceDescription = choice.Description
+				break
+			}
+		}
+	}
+
+	return resp, nil
+}
+
+func (mw *MultiWallet) agenda(agendas []chaincfg.ConsensusDeployment, agendaID string) *chaincfg.ConsensusDeployment {
+	for _, agenda := range agendas {
+		if agenda.Vote.Id == agendaID {
+			return &agenda
+		}
+	}
+
+	return nil
+}
+
+// SetVoteChoice sets a voting choice for the specified agenda using the
+// specified ticket hash controlled by a VSP.
+//
+// If a ticket hash isn't provided, the vote choice is set for all tickets
+// controlled by the set VSP.
+func (mw *MultiWallet) SetVoteChoice(walletID int, vspHost string, vspPubKey []byte, agendaID, choiceID, hash string, passphrase []byte) error {
+	wallet := mw.WalletWithID(walletID)
+	if wallet == nil {
+		return errors.New(ErrNotExist)
+	}
+	wal := wallet.Internal()
+
+	if vspHost == "" {
+		return fmt.Errorf("request requires a vspHost but no vspHost was provided")
+	}
+
+	// Test to ensure the passphrase is correct, and the voting is authorized.
+	err := wallet.UnlockWallet(passphrase)
+	if err != nil {
+		return translateError(err)
+	}
+	defer wallet.LockWallet()
+
+	var ticketHash *chainhash.Hash
+	if hash != "" {
+		hash, err := chainhash.NewHashFromStr(hash)
+		if err != nil {
+			return fmt.Errorf("inavlid hash: %w", err)
+		}
+		ticketHash = hash
+	}
+
+	choice := w.AgendaChoice{
+		AgendaID: agendaID,
+		ChoiceID: choiceID,
+	}
+
+	ctx := wallet.shutdownContext()
+	_, err = wal.SetAgendaChoices(ctx, ticketHash, choice)
```

Comment:
Want to note that this may succeed but updating the vote choices with the vsp may fail. If that happens, the GetVoteChoices method will return this choice saved to the wallet as the current choice whereas that choice is only saved locally and is not sent to the vsp. That can be grossly misleading.

Reply:
if voting with the vsp fails, the user would be notified with an error method, and i believe there should be a method that does reprocessing of the tickets, because saving the choices locally is what is done upstream as well, but if i'm wrong what do you suggest we do?

Reply:
exactly that was my major concern, it deviating from upstream, right now i probably need to update my decrediton or so, as trying to vote on agendas doesn't work for me

but upstream might not have thought about this, so we don't necessarily need to copy them, we can revert the voting preference as a temp solution until we have better information?